server/py/services/api/crud/secrets.py

```diff
@@ -419,6 +422,112 @@ def get_project_secret(
+        )
+
+        # TODO: move init iguazio_client
+        iguazio_client = framework.utils.clients.iguazio.v4.Client()
+        iguazio_client.refresh_access_tokens(secret_tokens)
+
+        created_tokens, updated_tokens, skipped_tokens = [], [], []
+
+        for token_name, decoded_token in valid_tokens.items():
+            action = self.secrets_provider.create_or_update_user_token_secret(
+                username=authenticated_username,
+                token_name=token_name,
+                token=decoded_token["raw_token"],
+                expiration=decoded_token["exp"],
+            )
+            if action == mlrun.common.schemas.SecretEventActions.created:
+                created_tokens.append(token_name)
+            elif action == mlrun.common.schemas.SecretEventActions.updated:
+                updated_tokens.append(token_name)
+            elif action == mlrun.common.schemas.SecretEventActions.skipped:
+                skipped_tokens.append(token_name)
+
+        return mlrun.common.schemas.StoreSecretTokensResponse(
+            created_tokens=created_tokens,
+            updated_tokens=updated_tokens,
+            skipped_tokens=skipped_tokens,
+        )
+
+    def _validate_and_decode_offline_tokens(
+        self,
+        secret_tokens: list[mlrun.common.schemas.SecretToken],
+        expected_user_id: str,
+    ) -> dict[str, dict]:
+        seen_names = set()
+        result = {}
+
+        for token in secret_tokens:
+            self._validate_token_name(token.name, seen_names)
+            decoded = self._decode_and_verify_offline_token(
+                token.name, token.token, expected_user_id
+            )
+            decoded["raw_token"] = token.token  # Add original token string
+            result[token.name] = decoded
+
+        return result
+
+    def _validate_token_name(self, token_name: str, seen_names: set):
+        if not token_name or token_name in seen_names:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Invalid or duplicate token name '{token_name}' found in request payload"
+            )
+        seen_names.add(token_name)
+
+    def _decode_and_verify_offline_token(
+        self, token_name: str, token: str, expected_user_id: str
+    ) -> dict:
+        decoded = self._decode_offline_token(token_name, token)
+
+        sub = decoded.get("sub")
+        exp = decoded.get("exp")
+
+        if not sub or not exp:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Offline token '{token_name}' is missing required claims (exp/sub)"
+            )
+
+        if sub != expected_user_id:
+            raise mlrun.errors.MLRunAccessDeniedError(
+                f"Offline token '{token_name}' does not belong to the authenticated user"
+            )
+
+        return decoded
+
+    def _decode_offline_token(self, token_name: str, token: str) -> dict:
+        try:
+            return jwt.decode(token, options={"verify_signature": False})
```

Comment:
Maybe add a comment here explaining that the token is already verified so there's no need to verify it here